base_classes/NXdata.nxdl.xml

```diff
@@ -115,6 +115,38 @@
 			the axes attribute can be omitted.
 		</doc>
 	</attribute>
+	<attribute name="default_slice">
+		<doc>
+			Which slice of data to show in a plot by default. This is useful especially for
+			datasets with more than 2 dimensions.
+
+			Should be an array of strings of length equal to the number of dimesions
+			in the data, with the following possible values:
+			* ".": All the data in this dimension should be included
+			* Integer: Only this slice should be used.
```

Comment:
Reading this again, I don't think it is completely clear what is meant here. An actual integer or a string containing only an integer? Since the type is NX_CHAR, I suppose it is the latter, but I still think it is a bit confusing. Maybe you could add an example with an integer as well?

Reply:
Hi, sorry for the long delay here.  Attributes like the proposed default_slice don't have a NeXus type, but you are right that you can't mix strings and integers in an array.  I've added the following as an example:

```
Alternate version using an integer would look like this (note 2 is a string)::
    @default_slice = [".", "2", ".", "."]
```

A little weird but I think it covers the requirements.

Reply:
The rendered docs say it has type `NX_CHAR`, I think this is just the default when you don't specify anything else.
![image](https://github.com/nexusformat/definitions/assets/73243774/636d11a7-1830-4a4d-a62f-08e6f727e543)

I think I would give a complete example or at least show that `channel` no longer contains strings.

Reply:
The length of the dimension over which you want to slice is `nC` which is a relatively small number. So why not use `auxiliary_signals` for this?

```
data:NXdata
   @signal = "difference"
   @auxiliary_signals = ["threshold_1", "threshold_2"]
   @axes = ["image_id", ".", "."]
   difference = unit[nP, i, j]
   threshold_1 = unit[nP, i, j]
   threshold_2 = unit[nP, i, j]
```

If for some reason you are forced to have a raw 4D dataset `data = uint[nP, nC, i, j]` then you can create three HDF5 virtual datasets 

```
threshold_1 = VDS(data[:, 0, :, :])
threshold_2 = VDS(data[:, 1, :, :])
difference = VDS(data[:, 2, :, :])
```

Reply:
Sorry for my very late reply. For us, the channels are comparable to an image with RGB channels, since they belong to the exact same exposure. One wouldn't save a separate image for the red, green and blue channel as well. Therefore, we don't want to split the dataset in multiple three-dimensional datasets. Splitting them would mean their logical connection is no longer visible and it is also no longer a given that the dimensions will match.

Regarding your suggestion with adding multiple VDSs, that would be a possibility if it also makes sense from a processing point of view. Our 4D dataset is a VDS already (the raw data is stored in multiple files) and I do not know if that has any unwanted consequences (for example longer loading times etc.). Do you by chance have experience with this?

Reply:
@soph-dec discussed in Telco today:

- In HDF5 user meeting today, it was presented that VDS adds a 5% overhead
- @PeterC-DLS suggests in the past for a different use-case that VDS was a read-back performance problem with a round-robin data saving framework which saves virtually interlaced frames stitched together in a VDS dataset, but he suspects this was addressed in recent HDF5 versions.  Indeed, @yayahjb note that VDS performance is very sensitive to HDF5 version.
- One note is that the proposed NXdetector_channel base class (#1252) provides support not mentioned here of adding additional metadata from NXdetector (such as flatfield or pixelmask) on a per channel basis.
- One correction: "Splitting them would mean their logical connection is no longer visible and it is also no longer a given that the dimensions will match."  The first part is true, but FYI `@auxillary_signal` specifies the additional signals are required to be the shape of the original signal.

Reply:
> Our 4D dataset is a VDS already (the raw data is stored in multiple files)

If it is already a virtual dataset, would it work for you to create a virtual dataset for each channel separately instead of creating one virtual dataset for everything? This would also match better with the https://github.com/nexusformat/definitions/pull/1252 mentioned before.

Reply:
The reason I'm hesitant about `default_slice` is our previous experience with `AXISNAME_indices`. No Nexus viewer that I know of supports it. The `default_slice` attribute will likely go down the same road: writers will create it but reader will ignore it. Ping @rayosborn

Reply:
I was looking at how we could add the 3D datasets without loosing the 4D dataset. That would then look something like this (note that we actually also want to use the `AXISNAME_indices` attribute for the `start_time` dataset which relates to the "image_id" axis):

```
data:NXdata
  @signal = "difference"
  @auxiliary_signals = ["threshold_1", "threshold_2"]
  @axes = ["image_id", ".", "."]
  @image_id_indices = 0
  @start_time_indices = 0
  difference = uint[nP, i, j]
  threshold_1 = uint[nP, i, j]
  threshold_2 = uint[nP, i, j]
  image_id = uint[nP]
  start_time = uint[nP]
  data = uint[nP, nC, i, j]
    @axes = ["image_id", "channel", ".", "."]
```

The problem I have now is that I can no longer specify the oder of the channels in the `data` dataset. The `channel` dataset in the NXdata group would vanish, but since the 4D data ist just another "DATA" field of the NXdata group, I can specify the axes attribute but no longer the corresponding `channel` dataset. Do you maybe have a suggestion how to solve that?
Also, I do not know how to specify that the `start_time` corresponds to the "image_id" axes of `data` as well.

I am open to other suggestions, but we really do not want to loose the 4D dataset.

Reply:
> * One note is that the proposed NXdetector_channel base class ([Detector channels #1252](https://github.com/nexusformat/definitions/pull/1252)) provides support not mentioned here of adding additional metadata from NXdetector (such as flatfield or pixelmask) on a per channel basis.

I'm sorry, I don't quite understand what you mean with this comment. Could you elaborate?

Reply:
If you have 3D datasets, why do you still need to 4D dataset? You could just do this:

```
data:NXdata
  @signal = "difference"
  @auxiliary_signals = ["threshold_1", "threshold_2"]
  @axes = ["image_id", "start_time"]
  @image_id_indices = 0
  @start_time_indices = 0
  difference = uint[nP, i, j]
  threshold_1 = uint[nP, i, j]
  threshold_2 = uint[nP, i, j]
  image_id = uint[nP]
  start_time = uint[nP]
```

So this NXdata group has three 3D signals ("difference" is the default).

As for coordinates: the first data dimension has two alternative coordinates (`image_id` and `start_time`) while the other two dimensions do not have coordinates.

FYI see [here](https://hdf5.gitlab-pages.esrf.fr/nexus/nxdata_axes/classes/base_classes/NXdata.html)  for an improved documentation on NXdata (https://github.com/nexusformat/definitions/pull/1213). Long story short: when `AXISNAME_indices` is provided, the order of `@axes` becomes irrelevant (it will just be used by readers to know which datasets contain coordinates).

Reply:
If you still need to 4D dataset you can have it of course. It just won't be part of the NXdata definition for the 3D signals.

Reply:
If I understand correctly, the data has to be written in a 4D dataset. How about this:

```
+-- entry
    |
    +-- data@NXdata
         |
         +-- difference -> /entry/instrument/detector/difference_channel/data
         +-- threshold_1 -> /entry/instrument/detector/threshold_1_channel/data
         +--  threshold_2 -> /entry/instrument/detector/threshold_2_channel/data
    +-- instrument@NXsignal
        |
        +-- detector@NXdetector
            |
            +-- difference_channel
            |      |
            |     +-- data(3D) -> VDS(/entry/instrument/detector/data[0,...])
            +-- threshold_1_channel
            |      |
            |      +-- data(3D) -> VDS(/entry/instrument/detector/data[1,...])
            +-- threshold_2_channel
            |     |
            |      +-- data(3D) -> VDS(/entry/instrument/detector/data[2,...])
            |
            +-- data (4D: detector dumps the results in here)
```

`NXdetector` has a recommended `data` field so that matches nicely. The new `NXdetector_channel` class could have a recommended `data` field as well for consistency. And finally the NXdata group on the top provided a default plot for this data without the need to change NXdata.

Reply:
I have thought about the possibility to put the data into the detector/detector_channels, but I still do not see how I can define the order of the channels in the 4D dataset or add the `start_time` axis to it with this solution.

Reply:
Regarding the 4D dataset: We want to use that because for our use case it is the correct way to represent the data. Adding the 3D datasets would only be a workaround for the limitations of the current NXdata definition.

> FYI see [here](https://hdf5.gitlab-pages.esrf.fr/nexus/nxdata_axes/classes/base_classes/NXdata.html) for an improved documentation on NXdata (#1213).

But I looked at the new documentation and noticed this:
![image](https://github.com/nexusformat/definitions/assets/73243774/2bc237ae-2762-464a-8b21-36b76aef385a)

Ranks up to 32 are supported? What was the idea for plotting data with higher ranks than 3? Maybe all of this has already been solved before?

In addition, through a link in the new documentation I found [this](https://www.nexusformat.org/NIAC2018Minutes.html#nxdata-plottype--attribute), stating

> NeXus is about how to find and annotate the data to be plotted but not to describe how the data is to be plotted.

Does the proposed `default_slice` fall into "how the data has to be plotted"?

I feel like at this point we might come to a conclusion faster if we have a call and discuss this directly. Would you be open to that? I'll send you an email so we might find a time.

Reply:
> Does the proposed default_slice fall into "how the data has to be plotted"?

No I don't think so. It is more equivalent to what `@signal` vs. `@auxiliary_signals` does.

Reply:
> Ranks up to 32 are supported

The docs say

> The maximum rank is 32 for compatibility with backend file formats.

There is otherwise no reason to restrict the rank at all. A reader typically plots NXdata as a 1D plot (curve), 2D plot (image) or 3D plot (volume rendering). In addition it will provide slider to browse through the other dimensions,

Reply:
From the meeting with @phyy-nx @woutdenolf @soph-dec it is more clear why the request was made. Dectris could easily make 3D datasets or a 4D dataset with 3D virtual datasets. The request is more functional: it doesn't feel right to split up multi-channel images. This would be like creating a format for RGB images and each channel needs to be kept separately.

Still you could argue that the `signal-auxiliary_signals` approach works and it would not require a modification of `NXdata` and hence all `NXdata` reader software. However when you don't have 3 channels but douzens or even hundreds of channels, creating a dataset or virtual dataset for each channels will make the file structure unreadable. One example is the new proposal for the NXxas (https://github.com/nexusformat/definitions/pull/1293) where you essentially want to save 2D data with many columns and each column is a counter.

So for that reason I agree we need to support axes with string type. And since you don't have `@signal` vs. `@auxiliary_signals` we would need `@default_slice`.